subprojects/messaging/src/main/java/org/gradle/internal/serialize/ExceptionPlaceholder.java

```diff
@@ -142,20 +142,53 @@ public Object transform(Object obj) {
         return Collections.emptyList();
     }
 
-    private static List<ExceptionPlaceholder> convertToExceptionPlaceholderList(List<? extends Throwable> throwables, Transformer<ExceptionReplacingObjectOutputStream, OutputStream> objectOutputStreamCreator) {
+    private static List<ExceptionPlaceholder> convertToExceptionPlaceholderList(Throwable original, List<? extends Throwable> throwables, Transformer<ExceptionReplacingObjectOutputStream, OutputStream> objectOutputStreamCreator, boolean checkCycle) {
         if (throwables.isEmpty()) {
             return Collections.emptyList();
         } else if (throwables.size() == 1) {
-            return Collections.singletonList(new ExceptionPlaceholder(throwables.get(0), objectOutputStreamCreator));
+            Throwable cause = throwables.get(0);
+            if (checkCycle) {
+                if (hasCycle(original, cause)) {
+                    return Collections.emptyList();
+                }
+            }
+            return Collections.singletonList(new ExceptionPlaceholder(cause, objectOutputStreamCreator));
         } else {
             List<ExceptionPlaceholder> placeholders = new ArrayList<ExceptionPlaceholder>(throwables.size());
             for (Throwable cause : throwables) {
-                placeholders.add(new ExceptionPlaceholder(cause, objectOutputStreamCreator));
+                if (!checkCycle || !hasCycle(original, cause)) {
+                    placeholders.add(new ExceptionPlaceholder(cause, objectOutputStreamCreator));
+                }
             }
             return placeholders;
         }
     }
 
+    private static boolean hasCycle(Throwable original, Throwable source) {
+        Throwable current;
+        Throwable cause = source;
+
+        try {
+            do {
+                current = cause;
+                if (original == current) {
+                    return true;
+                }
+                List<? extends Throwable> suppressed = extractSuppressed(current);
+                if (!suppressed.isEmpty()) {
```

Comment:
Why check for empty before the loop? I think we can rely on JIT doing these things

Reply:
Not sure the JIT can prevent an iterator creation in that case tbh